subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/transform/MacroTransformation.java

```
@@ -18,42 +18,113 @@
-        final ClassCodeExpressionTransformer trn = new ClassCodeExpressionTransformer() {
+    protected GroovyCodeVisitor getTransformer(ASTNode[] nodes, final SourceUnit sourceUnit) {
+        // Macro methods should on a classpath of the compiler because we invoke them during the compilation
+        final ClassLoader classLoader = this.getClass().getClassLoader();
+        return new ClassCodeVisitorSupport() {
+
             @Override
             protected SourceUnit getSourceUnit() {
                 return sourceUnit;
             }
 
             @Override
-            public Expression transform(final Expression exp) {
-                if (exp instanceof ConstructorCallExpression) {
-                    MethodCallExpression call = exp.getNodeMetaData(MacroTransformation.class);
-                    if (call!=null) {
-                        return call;
+            public void visitMethodCallExpression(MethodCallExpression call) {
+                super.visitMethodCallExpression(call);
+
+                List<MethodNode> methods = MacroMethodsCache.get(classLoader).get(call.getMethodAsString());
+
+                if (methods == null) {
+                    // Not a macro call
+                    return;
+                }
+
+                List<Expression> callArguments = InvocationWriter.makeArgumentList(call.getArguments()).getExpressions();
+
+                ClassNode[] argumentsList = new ClassNode[callArguments.size()];
+
+                for (int i = 0; i < callArguments.size(); i++) {
+                    argumentsList[i] = ClassHelper.make(callArguments.get(i).getClass());
+                }
+
+                methods = StaticTypeCheckingSupport.chooseBestMethod(MACRO_CONTEXT_CLASS_NODE, methods, argumentsList);
+
+                for (MethodNode macroMethodNode : methods) {
+                    if (!(macroMethodNode instanceof ExtensionMethodNode)) {
+                        // TODO is it even possible?
```

If you don't know if it's possible, it's probably better to throw an error. Then if the case appears we know we need to fix it.